### ISSUE TITLE
LBR Source/Target Address Support

### ISSUE THREAD

**Issue by fjricci** [iovisor/bcc#2491]
Hi! I found some old issues (ie #1234 from 2017) which mention that LBR support was on the BPF TODO list. Has this work been completed? If not, is there any tracking/discussion about it that I can follow or contribute to?

I can write a bpf program that watches for `BR_INST_RETIRED.NEAR_TAKEN` events, but I don't have a good way to actually access the source/target address ring buffer (doesn't look like I can read the `MSR_LASTBRANCH_*` registers directly from BPF, and `bpf_perf_event_read_value` just returns the pmu counter value, not the source/address values).

**Comment by yonghong-song** [comment]
No. This is not completed. So welcome to contribute.
Unfortunately, except this task, we did not have many other discussions on this topic.

No, you cannot directly read hard registers (MSR_LASTBRANCH_*) in bpf programs. As you correctly stated it, bpf_perf_event_read_value only returns pmu counters and elpsed times.

Currently, kernel traverses user space assuming fp available. Some work needs to be done in kernel to get LBR stack if LBR is configured and use it to get addresses.

**Comment by ygdyang** [comment]
hi，is the LBR support in BPF has been completed yet? @yonghong-song @fjricci 
I have some trouble when using `get_stackid`, because the frame pointer is omitted, I think the LBR support may give me some help.

**Comment by lingchengfenzi** [comment]
> hi，is the LBR support in BPF has been completed yet? @yonghong-song @fjricci I have some trouble when using `get_stackid`, because the frame pointer is omitted, I think the LBR support may give me some help.

What did you do then? I have the same trouble.

**Comment by anakryiko** [comment]
yes it is supported now, see `bpf_get_branch_snapshot()` helper. E.g., retsnoop tool is [utilizing it](https://github.com/anakryiko/retsnoop/blob/f82e0f83c7d385cc15d4b7f62ebc80d610d394f7/src/mass_attach.bpf.c#L60-L69), if you need an example.
